### PR TITLE
[FLINK-6830] Add StatefulJobSavepointFrom13MigrationITCase

### PR DESCRIPTION
**Pull request by tzulitai**
With [FLINK-6763](https://issues.apache.org/jira/browse/FLINK-6763) and [FLINK-6764](https://issues.apache.org/jira/browse/FLINK-6764) we'll need to change the serialization formats between 1.3.0 and 1.3.x.

This PR adds the ITCase for the savepoint migration so that those format changes will be properly guarded for backwards compatibility.

The savepoint binary files of this PR were taken using the current `release-1.3` branch.

### CODE REVIEW

**Comment by zentol** [comment]
Can we also include the tests for the restore with topology changes in flink-tests?

org.apache.flink.test.start.operator.restore *

**Comment by aljoscha** [comment]
The changes look good! @zentol's suggestion also seems wise. 👌

**Comment by tzulitai** [comment]
Thanks for the reviews @zentol @aljoscha! I'll add the topology change ITTests also and then merge this.

**Comment by aljoscha** [comment]
@tzulitai While you're on it, what would you think about also porting the `*From12MigrationTest` tests?

**Comment by tzulitai** [comment]
@aljoscha +1 to that, I'm actually working on those also already :-D

**Comment by tzulitai** [comment]
f228081 is for porting the topology change on restore migration tests.

The rest 233c617 to e619b98 is for ports of each individual migration test with the pattern `*From12MigrationTest`. They're also refactored to deduplicate code across 1.1, 1.2, 1.3 migration tests.

Note that for the CEP migration tests in 233c617, some tests for migration from 1.3 is currently ignored because they do not pass. May have bumped into a bug for the CEP library that does not allow it to restore from 1.3.x to 1.3.x 😅

File: flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/unkeyed/ChainBreakTest.java [hunk 1]
```diff
@@ -31,8 +36,22 @@
 /**
  * Verifies that the state of all operators is restored if a topology change breaks up a chain.
  */
+@RunWith(Parameterized.class)
 public class ChainBreakTest extends AbstractNonKeyedOperatorRestoreTestBase {
 
+	private final String savepointPath;
+
+	@Parameterized.Parameters(name = "Migrate Savepoint: {0}")
+	public static Collection<String> parameters () {
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/unkeyed/ChainBreakTest.java]: is it not possible to move this into the abstract super class?

**Comment by zentol** [comment]
topology tests look good.

File: flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/unkeyed/ChainBreakTest.java (hunk 1 above)
Comment (tzulitai) [inline flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/unkeyed/ChainBreakTest.java]: Will change!

**Comment by aljoscha** [comment]
Why are the CEP "from12" snapshots also changed?

**Comment by aljoscha** [comment]
@kl0u Could you please look into the CEP tests that failing to restore from 1.3 to 1.3?

**Comment by kl0u** [comment]
@aljoscha yes we have already discussed offline with @tzulitai .

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorMigrationTest.java [hunk 2]
```diff
@@ -149,9 +169,16 @@ public void testRestoreSessionWindowsWithCountTrigger() throws Exception {
 				new KeyedOneInputStreamOperatorTestHarness<>(operator, new TupleKeySelector(), BasicTypeInfo.STRING_TYPE_INFO);
 
 		testHarness.setup();
-		testHarness.initializeState(
-				OperatorSnapshotUtil.readStateHandle(
-						OperatorSnapshotUtil.getResourceFilename("win-op-migration-test-session-with-stateful-trigger-flink1.2-snapshot")));
+
+		String savepointFile = "win-op-migration-test-session-with-stateful-trigger-flink" + testMigrateVersion + "-snapshot";
+		if (testMigrateVersion.equals("1.1")) {
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorMigrationTest.java]: I think you could refactor this piece of code into a method, something like:
```
restoreFromSnapshot(TestHarness, SnapshotPath, MigrationVersion)
```

What do you think?

**Comment by aljoscha** [comment]
I like these changes a lot! Especially the refactoring of using the same test for all versions. (I'm sure @zentol will also like that 😉 )

Had some inline comments.

**Comment by kl0u** [comment]
Hi @tzulitai ! Thanks for testing also this. 

The problem with the `CEP` failing test for `1.3` while passing for `1.2` is that the semantics of the `followedBy()` changed between `1.2` and `1.3`, so the expected results cannot be the same. 

In `1.2`, `followedBy()` implied non-deterministic relaxed contiguity while in `1.3` it is just relaxed contiguity. So for `1.3`, either change the expected result to contain only one match with the `foo1` as middle element, or change the pattern to have `followedByAny(middle)` (and in this case you have to take the savepoint again).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorMigrationTest.java (hunk 2 above)
Comment (tzulitai) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorMigrationTest.java]: That makes sense! I'll add the method to some `MigrationTestUtil` perhaps.

**Comment by tzulitai** [comment]
Thanks @kl0u, your suggestions fixes the problem. Would it also be possible to include document the contiguity semantic changes in #4041?

**Comment by kl0u** [comment]
@tzulitai yes I will ;)

**Comment by tzulitai** [comment]
Addressed @aljoscha @kl0u @zentol's comments.

Also rebased on `master` for another Travis run. If green, will merge by the end of today to `master` and `release-1.3` if there's no other objections!

**Comment by tzulitai** [comment]
Merging ...

**Pull request closed without merging**